### DISCUSSION TITLE
Cannot read properties of null (reading 'registered') - Next.js

### DISCUSSION THREAD

**Discussion by Yeonny0723** [chakra-ui/chakra-ui#8804] [General]
#### Custom Library
- "@chakra-ui/react": "^2.8.2"
- "react": "^18.2.0"
-  "vite": "^5.1.4"
-  "typescript": "^5.2.2"

#### NextJS Project
- "next": "^14.2.7"
-  "react": "^18.3.1"
-  "typescript": "^5"
- "turbo": "^2.0.14"

<br/>

## Background 

Hi, I created a library using React + Chakra UI and tried to implement it within my NextJS project. 
The library includes the React component as well as some custom hooks, they all got imported well on the other React project. 
But when importing components within the NextJS, I bumped into an error called "⨯ TypeError: Cannot read properties of null (reading 'registered')" on the terminal server whereas my source doesn't contain any snippet of reading "registered". (I Guess .registered got called on the building stage,)

![image (1)](https://github.com/user-attachments/assets/65ff524b-174a-47d4-b8c7-ae84f1bf3806)

<br/>

## What I tried...
**(1) Specified the component as a client-only component using the "use client" directive -> Failed ❌**
I assumed that NextJS somehow includes the client component into a server-side rendering to optimize on their own though it is supposed to be a client only.

**(2) Enabled ssr on Vite library mode to support SSR of my library -> Failed ❌**
```javascript
// vite.config.js
export default defineConfig({
  plugins: [
    react(),
    dts({
      insertTypesEntry: true,
    }),
  ],
  build: {
    ssr: true,
```

**(3) Used dynamic import -> Succeed ⭕** 
However, it is only effective at the individual component level not for the entire library. 
```typescript
import dynamic from "next/dynamic";
const Avatar = dynamic(() => import("@library-name").then(mod => mod.Avatar), { ssr: false });
```

<br/>

## Related Issues
[react-select: Cannot read properties of null (reading 'registered')](https://github.com/JedWatson/react-select/issues/5555)
[Cannot read properties of null (reading 'registered') - Next.js](https://github.com/chakra-ui/chakra-ui/discussions/8521)

Your support would be greatly helpful!

### Link to Reproduction

Not provided

### Steps to reproduce

Not provided

### Chakra UI Version

2.8.2

### Browser

Chrome

### Operating System

- [ ] macOS
- [X] Windows
- [ ] Linux

**Comment by Yeonny0723** [comment]
Problem solved !!

You might need a custom provider to implemet a new library republished using Chakra UI (it uses useContext internally).
Just dynamically import a provider only and wrap the top level client component, it would generates no error !!

```typescript
import { Avatar } from "@libraryName";

const CustomProvider= dynamic(() => import("@libraryName").then(mod => mod.CustomProvider), {
  ssr: false,
}); 

export default function ClientComponent() {
  return (
    <CustomProvider>
      <Avatar  />
    </CustomProvider>
  );
}


```

**Comment by segunadebayo** [comment]
Duplicate of https://github.com/chakra-ui/chakra-ui/discussions/8521

Same Emotion `registered` error when a Vite library re-exports Chakra into Next.js. The workaround you posted (dynamic-import the provider, `ssr: false`) belongs on that thread.
